content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+created by ```alloc``` are the original pointers into the zeroth index of memory and are treated like the "Owners" of that memory. It is these pointers that need to be freed when the scope of their region ends. Pointers created by ```id```, ```load```, and ```ptr_add``` are merely memory borrowers with views into heap memory. They still need to be statically checked that they aren't being used after the region scope ends but do not need a corresponding ```free``` instruction like memory owners do. What is important about this typing is that both pointer types can exist within a program depending on whether you want the region to be inferred or explicitly stated. Whether a pointer owns its memory is trivial to decide and it is not expected that programmers included this annotation. As will be touched on in the evaluation section, this opt-in use of explicit regions makes converting valid programs to this extension simple.
+
+## Current Limitations on Programs
+
+Outside of the type system additions, there are a few restriction put onto Bril Programs for them to be valid for this extension:
+
+- Any variable that owns it's memory(returned from ```alloc```) can not be overwritten during the course of the program. This is to ensure that a program doesn't make multiple allocations using the same variable name but whereas the static analysis can only free the variables once.
+- Currently, programmers cannot declare smaller regions within the function scope. Currently, each function has a region and each pointer argument has it's own region. The next step would be to allow the programmer to create their own scoped regions. This was set asider during the project proposal process but would be interesting further work.
+- As a consequence of the first two, a programmer cannot currently allocate memory in a loop. The ideal way to do this would to have the loop body be one region so that at the end of each loop iteration any memory allocated during that iteration is freed. I attempted to infer small regions just for this use case and found it to be both difficult infer and confusing to use. It can be very unclear where the analysis has chosen to start and end a region which makes for a pour experience when the programmer can't tell when a memory location goes out of scope.
+- Every Owning pointer must dominate the exit block(s) of it's region. In this way, you cannot conditionally allocate or free memory in a Bril Program. A given allocation of memory must exist in the same variable no matter which block it exits the region from. A borrowed pointer can be freely used as long as it is invalidated once the region scope ends.
+- The programmer can specify when two arguments to a function should have the same region but they are unable to declare one region as nested within another. This would allow finer grain control over what input regions are acceptable for a function can instead the coarseness of, "Either they are the same region or they aren't".
```

Comment:
Showing how this attribute of function arguments is expressed in the language would be helpful.